tests/unit/test_charm.py

```diff
@@ -0,0 +1,334 @@
+        mock_rm.reset_mock()
+
+        # update to 10 VMs with 4 cpu and 7GiB memory
+        harness.update_config({"virtual-machines": 10, "vm-cpu": 4})
+        harness.charm.on.reconcile_runners.emit()
+        token = harness.charm.service_token
+        rm.assert_called_with(
+            "github-runner",
+            "0",
+            RunnerManagerConfig(
+                path=GitHubRepo(owner="mockorg", repo="repo"),
+                token="mocktoken",
+                image="jammy",
+                service_token=token,
+            ),
+            proxies={},
+        )
+        mock_rm.reconcile.assert_called_with(
+            10, VirtualMachineResources(cpu=4, memory="7GiB", disk="10GiB")
+        )
+        mock_rm.reset_mock()
+
+    @patch("charm.RunnerManager")
+    @patch("pathlib.Path.write_text")
+    @patch("subprocess.run")
+    def test_on_stop(self, run, wt, rm):
+        rm.return_value = mock_rm = MagicMock()
+        harness = Harness(GithubRunnerCharm)
+        harness.update_config({"path": "mockorg/repo", "token": "mocktoken"})
+        harness.begin()
+        harness.charm.on.stop.emit()
+        mock_rm.flush.assert_called()
+
+    @patch("pathlib.Path.write_text")
+    @patch("subprocess.run")
+    def test_get_runner_manager(self, run, wt):
+        harness = Harness(GithubRunnerCharm)
+        harness.begin()
+
+        # Get runner manager via input.
+        assert harness.charm._get_runner_manager("mocktoken", "mockorg/repo") is not None
+
+        assert harness.charm._get_runner_manager() is None
+
+        # Get runner manager via config.
+        harness.update_config({"path": "mockorg/repo", "token": "mocktoken"})
+        assert harness.charm._get_runner_manager() is not None
+
+        # With invalid path.
+        assert harness.charm._get_runner_manager("mocktoken", "mock/invalid/path") is None
+
+    @patch("charm.RunnerManager")
+    @patch("pathlib.Path.write_text")
+    @patch("subprocess.run")
+    def test_on_install_failure(self, run, wt, rm):
+        """Test various error thrown during install."""
+
+        rm.return_value = mock_rm = MagicMock()
+        mock_rm.get_latest_runner_bin_url = mock_get_latest_runner_bin_url
+
+        harness = Harness(GithubRunnerCharm)
+        harness.update_config({"path": "mockorg/repo", "token": "mocktoken"})
+        harness.begin()
+
+        # Base case: no error thrown.
+        harness.charm.on.install.emit()
+        assert harness.charm.unit.status == ActiveStatus()
+
+        harness.charm._reconcile_runners = raise_runner_error
+        harness.charm.on.install.emit()
+        assert harness.charm.unit.status == MaintenanceStatus(
+            "Failed to start runners: mock error"
+        )
+
+        harness.charm._reconcile_runners = raise_error
+        harness.charm.on.install.emit()
+        assert harness.charm.unit.status == BlockedStatus("mock error")
```

Comment:
What kind of error is expected?

Reply:
Reconcile runners can have various type of errors such as LXD error, hardware related errors (out of memory). This test checks if these unexpected errors, sets the charm into BlockedStatus.